test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageBoundPropertyFilterTest.cs

```diff
@@ -0,0 +1,159 @@
+            });
+
+            var handler = new TestPageModel();
+            var pageContext = CreatePageContext();
+            var filter = CreateFilter(tempDataFactory, handler);
+            var handlerExecutingContext = CreateHandlerExecutingContext(handler, pageContext);
+
+            // Act
+            filter.OnPageHandlerExecuting(handlerExecutingContext);
+
+            // Assert
+            Assert.Equal("ValueFromRequest", handler.TempData);
+            Assert.Null(handler.ViewData);
+        }
+
+        [Fact]
+        public void OnResultExecuting_WritesValuesToCollections()
+        {
+            // Arrange
+            var tempDataFactory = GetTempDataFactory(new Dictionary<string, object>
+            {
+                { "TempData", "ValueFromRequest" },
+            });
+
+            var handler = new TestPageModel
+            {
+                TempData = "New-TempDataValue",
+                ViewData = "New-ViewDataValue",
+            };
+            var pageContext = CreatePageContext();
+            var filter = CreateFilter(tempDataFactory, handler);
+            var resultExecutingContext = CreateResultExecutingContext(handler, pageContext);
+            pageContext.HttpContext.Items[typeof(ViewDataDictionary)] = pageContext.ViewData;
+
+            // Act
+            filter.OnResultExecuting(resultExecutingContext);
+
+            // Assert
+            var tempData = (TempDataDictionary)tempDataFactory.GetTempData(pageContext.HttpContext);
+            Assert.Equal("New-TempDataValue", tempData["TempData"]);
+            Assert.Equal("New-ViewDataValue", pageContext.ViewData["ViewData"]);
+
+            // Ensure the key is retained
+            Assert.Equal(new[] { "TempData" }, tempData.RetainedKeys);
+        }
+
+        [Fact]
+        public void OnResultExecuting_SkipsWritingToDictionary_IfValueIsUnchaned()
+        {
+            // Arrange
+            var tempDataFactory = GetTempDataFactory(new Dictionary<string, object>
+            {
+                { "TempData", "ValueFromRequest" },
+            });
+
+            var handler = new TestPageModel
+            {
+                TempData = "ValueFromRequest",
+                ViewData = null,
+            };
+            var pageContext = CreatePageContext();
+            var filter = CreateFilter(tempDataFactory, handler);
+            var resultExecutingContext = CreateResultExecutingContext(handler, pageContext);
+            pageContext.HttpContext.Items[typeof(ViewDataDictionary)] = pageContext.ViewData;
+
+            // Act
+            filter.OnResultExecuting(resultExecutingContext);
+
+            // Assert
+            var tempData = (TempDataDictionary)tempDataFactory.GetTempData(pageContext.HttpContext);
+            Assert.Equal("ValueFromRequest", tempData["TempData"]);
+            Assert.Null(pageContext.ViewData["ViewData"]);
+
+            // Ensure the key is retained
+            Assert.Empty(tempData.RetainedKeys);
```

Comment:
The comment here is confusing. We ensure the key is retained by asserting that RetainedKeys is empty?

Reply:
Not changing this?